### PR TITLE
Issue 927: Add support for vertical coordinate transforms 

### LINKED ISSUE #927: Add support for vertical coordinate transforms
Add support to convert from native vertical coordinate (e.g. pressure, Z, hybrid, etc.) to meters. In particular, transforms for vertical coordinate systems used by MOM, ROMS, and CAM are needed

### LINKED ISSUE #1927: regular grids defined in NetCDF are treated as stretched grids
NetCDF has no means to explicitly describe a Regular grid (one where all cells are the same size). Hence, Regular grids are defined in NetCDF using 1D coordinate variables, with uniform spacing between each value. Currently VAPOR does not examine the values inside 1D coordinate variables and thus assumes that a 1D coordinate variable describes a "stretched" dimension. Thus when accessing variables with 1D coordinate variables from NetCDF these variables are represented using the StretchedGrid class (as opposed to the RegularGrid class). A number of algorithms in VAPOR (e.g. DVR) operate more efficiently on the RegularGrid class. 

Reported by @leighorf

### PR DESCRIPTION
**Pull request by clyne**
Closes #927 

This PR closes #1927, Add support for vertical coordinate transforms. Specifically, it adds for Atmosphere hybrid sigma pressure coordinates used by CAM (e.g. AdamPhilips/F2000C5_new120_025) and other data. 

This PR also generalizes the LayeredGrid class so that it supports non-uniformly sampled (stretched) horizontal coordinates. This was necessary to support the grid resulting from the conversion of Atmospheric hybrid coordinates to meters.

### CODE REVIEW

File: lib/vdc/GridHelper.cpp [hunk 1]
```diff
@@ -276,12 +278,11 @@ LayeredGrid *GridHelper::_make_grid_layered(
...
-	for (int i = 0; i<2; i++) { 
-		float *coords = blkvec[i+1];
-		hminu.push_back(coords[0]);
-		hmaxu.push_back(coords[dims[i]-1]);
-	}
+	vector <double> xcoords;
+	for (int i=0; i<dims[0]; i++) xcoords.push_back(blkvec[1][i]);
+
+	vector <double> ycoords;
+	for (int i=0; i<dims[1]; i++) ycoords.push_back(blkvec[2][i]);
```
Comment (shaomeng) [inline lib/vdc/GridHelper.cpp:285]: Depending on if this piece of code is performance critical, you could call the reserve() method of the vector before pushing back a bunch of values.

File: lib/vdc/LayeredGrid.cpp [hunk 2]
```diff
@@ -129,98 +123,148 @@ void LayeredGrid::GetBoundingBox(
...
+			//
+			float zk = 
+				_zrg.AccessIJK(iv[0], jv[0], kk) * lambda[0] +
+				_zrg.AccessIJK(iv[1], jv[1], kk) * lambda[1] +
+				_zrg.AccessIJK(iv[2], jv[2], kk) * lambda[2];
+
+				zcoords.push_back(zk);
+		}
+
+		if (! Wasp::BinarySearchRange(zcoords, coords[2], indices[2])) return(false);
```
Comment (shaomeng) [inline lib/vdc/LayeredGrid.cpp:208]: I believe it's better readability and maintenance if you call one of the std binary search functions. That is, std::lower_bound, std::upper_bound, or std::binary_search.

**Review by shaomeng — APPROVED** [review]
A comment in general, std::vector seems to be used pretty intensively still. Given that we just had a campaign to use std::array in applicable scenarios, I wonder if some of the std::vectors could be converted to be std::arrays ?

File: lib/vdc/DerivedVar.cpp [hunk 3]
```diff
@@ -2899,3 +2899,390 @@ void DerivedCoordVarStandardOceanSCoordinate::compute_g2(
...
+    for (size_t j=0; j<max[1]-min[1]+1; j++) {
+    for (size_t i=0; i<max[0]-min[0]+1; i++) {
+
+		// We are deriving coordinate values from data values, so missing
+		// values may be present
+		//
+		float l_ps = ps[j*rDims[0]+i];
+		if (l_ps == _psVarMV) l_ps = 0;
+
+		float pressure = (a[k]*p0) + (b[k]*l_ps);
```
Comment (StasJ) [inline lib/vdc/DerivedVar.cpp:3241]: `compute_a` and `compute_ap` are identical other than the multiplication by `p0` here. You could just have one function with `p0` as an optional parameter initialized to `1`.

File: lib/vdc/GridHelper.cpp (hunk 1 above)
Comment (clyne) [reply lib/vdc/GridHelper.cpp:285]: Yup. Thanks. But the constructors in general were not designed to be performant. If the need ever arrises lots of changes will be required :-)

File: lib/vdc/DerivedVar.cpp (hunk 3 above)
Comment (clyne) [reply lib/vdc/DerivedVar.cpp:3236]: Doh! Good catch. Fixed.

File: lib/vdc/LayeredGrid.cpp (hunk 2 above)
Comment (clyne) [reply lib/vdc/LayeredGrid.cpp:208]: Ha ha. I did this, but the implementation with lower_bound/upper_bound is actually more code than what I had. Oh well. I've committed it.

**Review by StasJ — APPROVED** [review]

File: lib/vdc/LayeredGrid.cpp (hunk 2 above)
Comment (shaomeng) [reply lib/vdc/LayeredGrid.cpp:208]: Well, my understanding is that a user (e.g. a LayeredGrid) directly uses binary search functionality provided by std::lower_bound or std::upper_bound, but not use a wrapper (e.g. wasp::BinarySearchRange). Or in other words, if no new code uses wasp::BinarySearchRange() then that wrapper can go away after a few release cycles. Anyway, it's a minor issue.

**Pull request merged**